### PR TITLE
Migrated the storage logic from file to kvstore

### PR DESCRIPTION
**Pull request by hardikhdholariya**
- Migrated the storage logic from file to kvstore
- Updated addentries, getdevices, cleanup operations
NOTE: merge operation is not updated.

### CODE REVIEW

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py [hunk 1]
```diff
@@ -13,34 +12,100 @@
...
+import json
+import cs_utils
+
+from splunk import rest
 
 
-MAX_TIME_EPOCH = 2147483647   # Tue Jan 19 2038 03:14:07
+MAX_TIME_EPOCH = 2147483647  # Tue Jan 19 2038 03:14:07
+
+CY_HOSTNAME_POSTFIXES_MACRO = "cs_device_inventory_hostname_postfixes"
```
Comment (VatsalJagani) [inline cyences_app_for_splunk/bin/device_inventory_v2_util.py:27]: I don't think this is required here as we are passing it as class argument.

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py [hunk 2]
```diff
@@ -49,349 +114,425 @@ def _internal_check_list_field_format(self, field):
...
+        self.session_key = session_key
+        self.logger = logger
+        self.operation = operation
+        self.hostname_postfixes = [
+            element.strip() for element in hostname_postfixes if element.strip()
+        ]
+        self.modify_devices = []
+        self.devices = self.read_kvstore_lookup(DEVICE_INVENTORY_LOOKUP_COLLECTION)
+        if self.devices:
+            self._convert_str_to_dict()
```
Comment (VatsalJagani) [inline cyences_app_for_splunk/bin/device_inventory_v2_util.py:140]: I would suggest keeping this inside the read_kvstore_lookup function itself.
OR
use self.read_devices(<lookup>)
            self._read_kvstore_lookup(<lookup>)
            and _convert here

Comment (VatsalJagani) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py:140]: like update_kvstore_entries is handling it internally.

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py [hunk 3]
```diff
@@ -49,349 +114,425 @@ def _internal_check_list_field_format(self, field):
...
+                sessionKey=self.session_key,
+                raiseAllErrors=True,
+            )
 
+    def __enter__(self):
+        return self
 
-    def is_match(self, device_entry, hostname_postfixes=[]):
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.operation == "cleanup":
```
Comment (VatsalJagani) [inline cyences_app_for_splunk/bin/device_inventory_v2_util.py:224]: Keep two list:
* updated_devices
* deleted_devices

Perform operation on both lists on all the exits.
That way we don't need self.operation

Comment (VatsalJagani) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py:224]: Before updated_devices execution, remove entries from the updated_devices list that are also part of deleted_devices.

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py (hunk 2 above)
Comment (VatsalJagani) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py:140]: Keep the self.devices up to date.
* Update self.devices when updated. And add the key to modified_device_list. (While storing build the dict by using self.devices) -> For memory optimization and avoid conflicts
* Remove self.devices when deleted. And add the key to deleted_device_list.

**Review by VatsalJagani — CHANGES_REQUESTED** [review]

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py (hunk 1 above)
Comment (hardikhdholariya) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py:27]: We are passing the values of the hostname_postfixes from this macro.

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py (hunk 2 above)
Comment (hardikhdholariya) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py]: Done

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py (hunk 3 above)
Comment (hardikhdholariya) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py]: Done

File: cyences_app_for_splunk/bin/device_inventory_v2_util.py (hunk 1 above)
Comment (VatsalJagani) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py:27]: Yes, but we are getting it as an argument. It is defined in two python files.

Comment (hardikhdholariya) [reply cyences_app_for_splunk/bin/device_inventory_v2_util.py]: It's my bad. removed unused variable.

**Review by VatsalJagani — APPROVED** [review]

**Pull request merged**